zio-query/shared/src/main/scala/zio/query/ZQuery.scala

```diff
@@ -379,17 +393,24 @@ final class ZQuery[-R, +E, +A] private (private val step: ZIO[(R, QueryContext),
    * automatically be batched.
    */
   final def zipWithPar[R1 <: R, E1 >: E, B, C](that: ZQuery[R1, E1, B])(f: (A, B) => C): ZQuery[R1, E1, C] =
-    ZQuery {
-      self.step.zipWithPar(that.step) {
-        case (Result.Blocked(br1, c1), Result.Blocked(br2, c2)) => Result.blocked(br1 && br2, c1.zipWithPar(c2)(f))
-        case (Result.Blocked(br, c), Result.Done(b))            => Result.blocked(br, c.map(a => f(a, b)))
-        case (Result.Done(a), Result.Blocked(br, c))            => Result.blocked(br, c.map(b => f(a, b)))
-        case (Result.Done(a), Result.Done(b))                   => Result.done(f(a, b))
-        case (Result.Fail(e1), Result.Fail(e2))                 => Result.fail(Cause.Both(e1, e2))
-        case (Result.Fail(e), _)                                => Result.fail(e)
-        case (_, Result.Fail(e))                                => Result.fail(e)
+    new ZQuery(cb =>
+      Promise.make[Nothing, Result[R, E, A]].flatMap { left =>
```

Comment:
We still need to incur the overhead of creating and awaiting two promises here. One thing we can probably do is create an optimized version of `foreachPar` that starts all the queries, stores the results in an array, and then aggregates all the results in a single pass.